Closes #15796 (building PDF docs of matplotlib via LaTeX)

jfbu opened the pull request:
## PR Summary

Make build of matplotlib documentation to PDF possible.

The #15796 issue is caused by the fact that `latex_elements` dict got redefined near end of `conf.py` hence the part about using LaTeX package enumitem and its `\setlistdepth` got lost. But anyway since Sphinx 1.5 there is `'maxlistdepth'` key, so use that.

Another problem arose related to `expdlist` very old LaTeX package. I realized it broke the build <s>with recent LaTeX</s> (when a description environment was found in a longtable cell). 

edited: this also breaks with LaTeX as old as 2012. Because I did not recall having seen the issue in 2018 when trying out building scipy's PDF documentation (but perhaps I had already patched `expdlist`), and as my laptop only has TeXLive 2018 and 2019 I erroneously thought the issue arose from LaTeX upstream change of array package. But I see now `expdlist` (not updated since 1999...) breaks LaTeX (even from 2012) when a description list ends up in a table cell. It is not recent issue.

<s>The cause is probably the upstream LaTeX changes in array package which last year invalidated tabu package. </s>

For time being I simply commented out usage of `expdlist`. Turns out that a patch of mine which got merged into scipy (also for matters of building PDF) in April 2018 fixes the <s>more recent</s> breakage caused by `expdlist`. As experimenting with builds of matplotlib (close to 3000 pages) is time costly, I did not much investigate if it would be worthwile using `expdlist` as patched in my scipy PR. (see commit message).

Finally, matplotlib documentation of "Mathtext" uses tons of Unicode math-like characters in plain text, and DejaVu Serif was lacking 134 of them. I thus configure the build to use XITS and it turns out it provided all glyphs. It comes with TeXLive TeX distributions of recent years. I used syntax which should work whether compiling on Mac OS or Linux and probably Windows.

As latex engine is xelatex, it is not easy without additional mark-up to use both OpenType fonts and traditional TeX fonts, so I removed the `\usepackage{txfonts}` and used FreeSans and FreeMono for the sans and mono famillies. This means though that the math symbols will be the Knuth Computer Modern ones. I left commented out two lines loading `unicode-math` and the OpenType Math font `XITS Math` which would fix that aspect. Perhaps they should be activated ?


## PR Checklist

*almost* None of this applies because it is simply a modification of `doc/conf.py`...

Arrgh
```
$ flake8 conf.py
conf.py:4:80: E501 line too long (80 > 79 characters)
conf.py:7:80: E501 line too long (82 > 79 characters)
conf.py:93:1: E402 module level import not at top of file
conf.py:95:1: E402 module level import not at top of file
conf.py:180:80: E501 line too long (83 > 79 characters)
conf.py:297:80: E501 line too long (81 > 79 characters)
```
but... none of this comes from my patch as it is all after line number 300 :-)

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

Tested the PDF build with Sphinx 1.8.5 and 2.2.1 and Python 3.7.5 and matplotlib at 3d1a1bcc4

jfbu commented:
Forced pushed to reword a commit message. I had forgotten that in https://github.com/scipy/scipy/commit/c881fdea7a7ef7c518b7898cfd53b2880af06778 I had already found out that `expdlist` broke description lists ending up in table cells. Turns out that matplotlib docs do have such an instance (with master at 3d1a1bcc4). As I indicated in my original post, the scipy patch could also be used here, but It simply commented out usage of `expdlist` not having examined closely the difference in PDF output between doing this or incorporating same LaTeX fix as in my scipy commit.

jfbu commented:
on further thought and having experimented with building numpy's docs to PDF, and now remembering better my similar experience with building scipy's PDF docs, I think it would be better for coherence with these two other projects (who probably have many shared users with matplotlib project) to not comment out usage of `expdlist` LaTeX package (as I did in this PR) but rather to  fix it the same as I did in https://github.com/scipy/scipy/commit/c881fdea7a7ef7c518b7898cfd53b2880af06778 and copied over to https://github.com/numpy/numpy/pull/15028). It is only a matter of how parameter descriptions look like in PDF, i.e. here is a piece of matplotlib doc with this PR
![Capture d’écran 2019-12-03 à 23 14 36](https://user-images.githubusercontent.com/2589111/70094407-c0595300-1622-11ea-8352-2f31e6526d3e.png)
compared to how numpy's docs look like (with usage of `expdlist`):
![Capture d’écran 2019-12-03 à 23 14 52](https://user-images.githubusercontent.com/2589111/70094439-cea76f00-1622-11ea-9857-c7b4aea1c2ec.png)
Personally I would rather prefer yet another style, with the type description being on same line of label and description on new line....

jfbu commented:
I have pushed some more commits. Sorry about mishap with `latex_elements['pointsize']` which I moved to a place where `latex_elements` dict had not been initialized, then reverted.

- use `expdlist` with fixes. This is following what is done at SciPy and now also at NumPy (https://github.com/numpy/numpy/pull/15028)

Before:
![Capture d’écran 2019-12-04 à 22 10 26](https://user-images.githubusercontent.com/2589111/70183690-85205800-16e6-11ea-9570-1c5cde8f14ad.png)

After:
![Capture d’écran 2019-12-04 à 22 13 20](https://user-images.githubusercontent.com/2589111/70183703-8a7da280-16e6-11ea-8490-a64dd492cfc8.png)

- use `XITS Math` and `unicode-math` to get blackboard alphabet `\mathbb` accept `b, c, d` letters.

Before:
![Capture d’écran 2019-12-04 à 22 16 19](https://user-images.githubusercontent.com/2589111/70184564-5c995d80-16e8-11ea-81a5-ed8184e663bd.png)


After:
![Capture d’écran 2019-12-04 à 22 18 07](https://user-images.githubusercontent.com/2589111/70183778-b13bd900-16e6-11ea-876b-0b0832773ebc.png)

LaTeX installation must have `xelatex` installed:  on Ubuntu xenial, I think this means `texlive-xetex`, and `xindy` are needed.  Also `latexmk` is needed for `make latexpdf`.

And https://github.com/alif-type/xits fonts are needed. Ubuntu's [texlive-fonts-extra](https://packages.ubuntu.com/search?suite=xenial&arch=any&mode=exactfilename&searchon=contents&keywords=XITS-Regular.otf)  provided it at xenial already. And `texlive-fonts-extra` is on Ubuntu xenial a [requirement of Sphinx LaTeX](https://www.sphinx-doc.org/en/master/usage/builders/index.html#sphinx.builders.latex.LaTeXBuilder).

*nota bene*: while working on this I have noticed a problem with `'xelatex'` (in `smartquotes = False`) context with the `"` character. Unfortunately `'xelatex'` converts `"versions"` in latex file (as smart quotes are turned off, `"versions"` in reST sources makes it unmodified to LaTeX file) into `”versions”` in PDF output by default with fonts such as XITS or also GNU FreeFont which is the default font used by Sphinx 2.x with `xelatex`.

tacaswell commented:
Thank you for your work on this @jfbu ! 

How does this overlap with #15799 ?

Don't worry about force-pushing to your branch or making mistakes that you fix.

> But LaTeX life is desperately complicated and counter-intuitive in so many cases.

too true....

jfbu commented:
> 
> How does this overlap with #15799 ?


Ah sorry when I updated this one I had forgotten about #15799... and I incorporated here in my last update half of it, i.e. usage of ``\and`` mark-up (which fixes overflowing but will still have until Sphinx 2.3.0 the sphinx-doc/sphinx/issues/6876 problem).

Better to keep the two separated, but when I will update #15799 to only contain the fix of author display on title page, part of your reviewing work over there will be marked "outdated"... because it was on commits which will be only in this one soon. (my rationale into incorporating into #15799 the commits present here was to allow testing the PDF build, but if you trust the picture I posted over there it is not really needed....)


> 
> Don't worry about force-pushing to your branch or making mistakes that you fix.
> 

Ok, I will clean up a bit this branch. And add a commit backporting the method sphinx-doc/sphinx/pull/6888 fix of the double quote issue when using  `xelatex` under `smartquotes = False` regime. This will need no change with future Sphinx incorporating that fix, except that it will simply become superfluous.


> > But LaTeX life is desperately complicated and counter-intuitive in so many cases.
> 
> too true....

jfbu commented:
I have force pushed to reorganize the commit history and messages into logically consecutive steps.

QuLogic commented:
e01fe337d9e03f52c8ca65fdab28af4a108acc6e seems to have the wrong commit message (and could be merged with the following commit.) Otherwise, I think this looks reasonable, though I have yet to test it.

QuLogic commented:
I tried this out, but even after installing the `texlive-xits`, it fails to find the font:
```
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!
! fontspec error: "font-not-found"
! 
! The font "XITS-Regular" cannot be found.
! 
! See the fontspec documentation for further information.
! 
! For immediate help type H <return>.
!...............................................  
```
The file name is `xits-regular.otf`, and the font name is `XITS Regular`, so why is it looking for `XITS-Regular`?

QuLogic commented:
Ah, I see we have this in our pgf documentation; xelatex can't find fonts installed in texlive directories instead of globally. I worked around this by giving it the file name `xits-regular`.

However, it also failed to find `FreeSans.otf`. In Fedora, it's a ttf (probably will be otf in 32), but even changing `Extension      = .ttf,` didn't work. Since `FreeSans` is also the name of the font, dropping the `Extension` line entirely seems to have fixed that.

jfbu commented:
> Ah, I see we have this in our pgf documentation; xelatex can't find fonts installed in texlive directories instead of globally. I worked around this by giving it the file name `xits-regular`.

The situation is a wee'bit complicated with xelatex+fontspec for fonts, things have changed overtime.  On my mac os x, I have a special set-up with symlinks in $HOME/Library/Fonts to point to the TrueType and OpenType repertories in the TeXLive tree.  However I often forget I had long time ago set-up these symlinks and this may lead me to have a smoother ride than some users (on mac os). As per Linux I can test on Ubuntu but I have no Fedora handy.

On TeXLIve based distribution, if xelatex does not find the font in system/user locations it will then, it can still find it indeed by filename via the services of kpsewhich which comes with these distributions.

> [e01fe33](https://github.com/matplotlib/matplotlib/commit/e01fe337d9e03f52c8ca65fdab28af4a108acc6e) seems to have the wrong commit message (and could be merged with the following commit.) Otherwise, I think this looks reasonable, though I have yet to test it.

should I add another setting (using filename for XITS Math)  and then squash commits?

QuLogic commented:
It should definitely work with a default setup, not sure if an option would be needed.

jfbu commented:
Squashed and rebased to get a clear commit message.

I don't know what to do for XITS loading. Here is my local set-up:
```
$ ls /usr/local/texlive/2020/texmf-dist/fonts/opentype/public/xits
XITS-Bold.otf        XITS-Italic.otf      XITSMath-Bold.otf
XITS-BoldItalic.otf  XITS-Regular.otf     XITSMath-Regular.otf
```

On my mac os x, filenames ignore case. So I could use `xits-regular.otf` . I have no `.ttf` for XITS on my system. How do filenames look on Fedora systems ?

QuLogic commented:
The file names for XITS on Fedora are like `/usr/share/texlive/texmf-dist/fonts/opentype/public/xits/xits-regular.otf`.

QuLogic commented:
Any progress here? What needs to be done?

jkseppan commented:
I installed a clean mactex (mactex-20200407.pkg) from homebrew on Catalina and upgraded all packages with `sudo tlmgr update --all`. The build fails with

```
kpathsea: Running mktextfm XITS-Regular
/usr/local/texlive/2020/texmf-dist/web2c/mktexnam: Could not map source abbreviation X for XITS-Regular.
/usr/local/texlive/2020/texmf-dist/web2c/mktexnam: Need to update /usr/local/texlive/2020/texmf-dist/fonts/map/fontname/special.map?
mktextfm: Running mf-nowin -progname=mf \mode:=ljfour; mag:=1; ; nonstopmode; input XITS-Regular
This is METAFONT, Version 2.7182818 (TeX Live 2020) (preloaded base=mf)


kpathsea: Running mktexmf XITS-Regular
! I can't find file `XITS-Regular'.
<*> ...; mag:=1; ; nonstopmode; input XITS-Regular

Please type another input file name
! Emergency stop.
<*> ...; mag:=1; ; nonstopmode; input XITS-Regular

Transcript written on mfput.log.
grep: XITS-Regular.log: No such file or directory
mktextfm: `mf-nowin -progname=mf \mode:=ljfour; mag:=1; ; nonstopmode; input XITS-Regular' failed to make XITS-Regular.tfm.
kpathsea: Appending font creation commands to missfont.log.


! Package fontspec Error: The font "XITS-Regular" cannot be found.
```

I think xetex shouldn't need TeX font metrics, so it shouldn't be calling mktextfm.

jkseppan commented:
Editing the file like so allows the build to continue:

```tex
\setmainfont{XITS}[
Extension = .otf,
  UprightFont    = *-Regular,
  ItalicFont     = *-Italic,
  BoldFont       = *-Bold,
  BoldItalicFont = *-BoldItalic,
]
\setsansfont{FreeSans}[
Extension = .otf,
  UprightFont    = *,
  ItalicFont     = *Oblique,
  BoldFont       = *Bold,
  BoldItalicFont = *BoldOblique,
]
\setmonofont{FreeMono}[
Extension = .otf,
  UprightFont    = *,
  ItalicFont     = *Oblique,
  BoldFont       = *Bold,
  BoldItalicFont = *BoldOblique,
]
% needed for \mathbb (blackboard alphabet) to actually work
\usepackage{unicode-math}
\setmathfont{XITSMath-Regular}[
Extension = .otf
]
```

Would this work on Linux?

jkseppan commented:
Here's a Dockerfile for testing on Fedora:

```Dockerfile
FROM fedora:latest
RUN dnf install -y texlive-scheme-full
RUN latexmk -pdf -dvi- -ps- Matplotlib.tex
```

Simply building this in the `build/latex` directory after the `Matplotlib.tex` file has been generated should work. But the first RUN line installs over 4000 packages, so this is going to take a while...

jkseppan commented on doc/conf.py:
```diff
@@ -304,33 +304,92 @@ def _check_dependencies():
+# the Unicode codepoints needed for the section about Mathtext
+# "Writing mathematical expressions"
+latex_elements['fontpkg'] = r"""
+\setmainfont{XITS}[
```
```suggestion
\setmainfont{XITS}[
  Extension      = .otf,
```

jkseppan commented on doc/conf.py:
```diff
@@ -304,33 +304,92 @@ def _check_dependencies():
+  BoldFont       = *-Bold,
+  BoldItalicFont = *-BoldItalic,
+]
+\setsansfont{FreeSans}[
```
```suggestion
\setsansfont{FreeSans}[
  Extension      = .otf,
```

jkseppan commented on doc/conf.py:
```diff
@@ -304,33 +304,92 @@ def _check_dependencies():
+  BoldFont       = *Bold,
+  BoldItalicFont = *BoldOblique,
+]
+\setmonofont{FreeMono}[
```
```suggestion
\setmonofont{FreeMono}[
  Extension      = .otf,
```

jkseppan commented on doc/conf.py:
```diff
@@ -304,33 +304,92 @@ def _check_dependencies():
+]
+% needed for \mathbb (blackboard alphabet) to actually work
+\usepackage{unicode-math}
+\setmathfont{XITS Math}
```
```suggestion
\setmathfont{XITSMath-Regular}[
  Extension = .otf
]
```

QuLogic commented:
> It fails in makeindex

You need to install `texlive-xindy` for that, I believe.

QuLogic commented:
That won't work for Fedora, because it has `FreeSans.ttf`, not `.otf`.

QuLogic commented:
I rebased this and added additional fixes for new bugs in the above-linked PR.
